Updated file_io.rst to give a MVP for production (#132)
reflectivity/orsopy: docs/fileio_usage.rst

=====

The easiest way to use :py:mod:`orsopy.fileio` (the module of :py:mod:`orsopy` that includes file reading and writing) to produce **metadata-rich .ort reduced reflectometry files** involves integrating this into your data reduction workflow.
Early in the workflow, the :py:mod:`orsopy.fileio` should be imported and an empty :code:`orsopy.fileio.orso.Orso` header object (here we also import :py:mod:`numpy` which will be used later).

.. code-block:: python

    from orsopy import fileio
    from orsopy.fileio import (Reduction, Person, Orso, Experiment,
                               Sample, DataSource, Measurement, InstrumentSettings)
    import numpy as np
    import datetime

    header = fileio.orso.Orso.empty()

Having created the empty header object we can start to populate the appropriate components of it.
It is generally a good idea to populate the components as particular steps occur in the reduction process.
In this example we fill out the experiment details and then wrap them in the appropriate class.

.. code-block:: python

    title = 'my title'
    instrument = 'PLATYPUS'

>     start_date = datetime.date.today()  # This needs to be a datetime object

Reviewer: ```suggestion
    start_date = datetime.date.today()  # This needs to be an object of the type datetime
```

    probe = 'neutron'
    facility = 'ESS'
    proposalID = 999999

>     doi = f'10.5286/ISIS.E.RB{99999}'

Reviewer: Have you deliberately mixed ISIS, ESS and ANSTO?

Author: yes

    experiment = Experiment(title, instrument, start_date, probe, facility, proposalID, doi)

Full details of the different components that can be populated can be found in the `documentation`_ here or in the `file format specification`_.
Note that this specification includes information regarding the **required** and optional components to be included for a file to be considered a **valid** .ort file.

> What we show here is not the minimum needed to write an `.ort file, rather the minium to fill out all the major fields with information which is likely to be available at time of writing.

Reviewer: ```suggestion
What we show here is not the minimum needed to write an `.ort` file, rather the minium to fill out all the major fields with information which is likely to be available at time of writing.
```

We now write out the user and sample details, but note that the sample can be alot more descriptive than just a name.

.. code-block:: python

    user_name = 'Jos Cooper'
    user_affil = 'ESS'

    user = Person(user_name, user_affil)

    # Now fill the sample details
    sample_name = 'Silicon'

    sample = Sample(sample_name)

    # The measurement details needs some instrument settings assigned first
    angle = 0.3  # This can be a float, or a range, though currently not a list
    wavelength = (1.8, 8.8)  # Again a float or range

    instrument_settings = InstrumentSettings(angle, wavelength)

    # Now add the instrument settings to the names of the files used
    files = ['test_data.raw']  # A list of all raw files used reduced

    measurment = Measurement(instrument_settings, files)

We can now wrap together the user, experiment, sample and measurement information into a single object.

.. code-block:: python
    data_source_info = DataSource(user, experiment, sample, measurment)

> It is not possible to write a .ort file without defining the columns present in the dataset, in this example we will have four columns of data, namely q, R, dR and dq (the final column is a description of the resolution function).

Reviewer: ```suggestion
It is not possible to write an `.ort` file without defining the columns present in the dataset, in this example we will have four columns of data, namely q, R, dR and dq (the final column is a description of the resolution function).
```

Columns are defined as follows, using the :code:`orsopy.fileio.base.Column` and :code:`orsopy.fileio.base.ErrorColumn` class objects.

.. code-block:: python

    q_column = fileio.base.Column(name='Qz', unit='1/angstrom', physical_quantity='wavevector transfer')

>     # Note that the unit is not fixed to 1/angstrom by the standard, but currently the data loader interprets all values as 1/angstrom irrespective of what is stated here

Reviewer: Is this true?

Author: sadly yes. there is no unit processing that I could find in the data loader, and all analysis packages currently read data specified with 1/nanometer as 1/angstrom

Author: oh, but no!
I was just not using the right unit!!!

Author: `1/nm` not `1/nanometer`
